numpy/_core/src/common/slice.hpp

```diff
@@ -0,0 +1,241 @@
+    Iterator operator--(int)
+    { auto tmp = *this; --*this; return tmp; }
+
+    /// @name Comparison operators
+    /// @{
+    constexpr bool operator==(Pointer r) const
+    { return ptr_ == r; }
+    constexpr bool operator!=(Pointer r) const
+    { return ptr_ != r; }
+    constexpr bool operator<(Iterator r) const
+    { return ptr_ < r.ptr_; }
+    constexpr bool operator<=(Iterator r) const
+    { return ptr_ <= r.ptr_; }
+    constexpr bool operator>(Iterator r) const
+    { return ptr_ > r.ptr_; }
+    constexpr bool operator>=(Iterator r) const
+    { return ptr_ >= r.ptr_; }
+    constexpr bool operator==(Iterator r) const
+    { return ptr_ == r.ptr_; }
+    constexpr bool operator!=(Iterator r) const
+    { return ptr_ != r.ptr_; }
+    /// @}
+
+  private:
+    Pointer ptr_;
+    SSize byte_stride_;
+};
+} // namespace np::slice_helper
+
+/** One-dimensional subset of an array.
+ *
+ * This template class represents a one-dimensional subset of an array.
+ * It can be created from a dynamic array by specifying three parameters:
+ * `start` is a pointer to the beginning of the array,
+ * `size` is the total number of elements in the subset, and
+ * `stride` represents the distance between each successive array element
+ * to include in the subset between beginnings of successive array elements.
+ *
+ * @tparam T The type of the elements of the array.
+ */
+template<typename T>
+class Slice {
+  public:
+    /// Element type.
+    using ElementType = T;
+    /// @name Iterators types
+    /// @{
+    using Iterator = slice_helper::Iterator<T>;
+    using ConstIterator = slice_helper::Iterator<T, true>;
+    /// @}
+
+    /// @name Iterators types STD style
+    using iterator = slice_helper::Iterator<T>;
+    using const_iterator = slice_helper::Iterator<T, true>;
+    /// @}
+
+    /** Construct a slice from dynamic array.
+     *
+     * @param start  Pointer of the first element.
+     * @param len    Number of elements.
+     * @param stride Stride between array elements in the size of
+     *               of the array's element.
+     */
+    constexpr Slice(T *start, SSize len, SSize stride) noexcept
+        : begin_(Iterator(start, stride)), len_(len)
+    {}
+    /** Construct a slice from dynamic byte array.
+     *
+     * @param start  Pointer of the first element.
+     * @param len    Number of elements.
+     * @param stride Stride between array elements in bytes.
+     */
+    constexpr Slice(char *start, SSize len, SSize stride) noexcept
+        : begin_(Iterator(start, stride)), len_(len)
+    {}
+
+    /// @name Iterators
+    /// @{
+    constexpr Iterator Begin() noexcept
+    { return begin_; }
+    constexpr ConstIterator Begin() const noexcept
+    { return ConstIterator(begin_); }
+    constexpr Iterator End() noexcept
+    { return begin_ + len_; }
+    constexpr ConstIterator End() const noexcept
+    { return ConstIterator(begin_ + len_); }
+    /// @}
+
+    /// @name Iterators STD style
```

Comment:
lower-case style (std::) is required for modern C++ for loop iterates e.g. (for-each x:x) while the upper camel case is defined for compilation with Google's style since the whole patches within this pull request follow Google's coding style.

Reply:
OK, makes sense to stick to a style guide - it was just the "methods" that threw me off, because python decided that those look like variables rather than classes.